tutorials/backend/hasura-authentication/tutorial-site/content/integrations/casdoor.md

```diff
@@ -0,0 +1,149 @@
+- Accessing logs auditing and recording.
+- Alibaba Cloud, Tencent Cloud, Qiniu Cloud image CDN cloud storage.
+- Customizable registration, login, and password retrieval pages.
+- Casdoor supports integration with existing systems by db sync, so users can transition to Casdoor smoothly.
+- Casdoor supports mainstream databases: MySQL, PostgreSQL, SQL Server, etc., and supports the extension of new
+  databases with plugins.
+
+In this section, you will learn how to integrate Casdoor with Hasura.
+
+## Deploy Casdoor
+
+Firstly, you should deploy a Casdoor instance.
+
+You can refer to the Casdoor official documentation for the
+[Server Installation](https://casdoor.org/docs/basic/server-installation).
+
+After a successful deployment, you need to ensure:
+
+- The Casdoor server is successfully running on **http://localhost:8000**.
+- Open your favorite browser and visit **http://localhost:7001**, you will see the login page of Casdoor.
+- Input `admin` and `123` to test login functionality is working fine.
+
+Then you can quickly implement a Casdoor-based login page in your own app with the following steps.
+
+## Configure Casdoor application
+
+1. Create or use an existing Casdoor application.
+2. Add a redirect url: `http://CASDOOR_HOSTNAME/login`
+   ![Casdoor Application Setting](https://github.com/RanTao123/image/blob/main/Casdoor%20Application%20Setting.png?raw=true)
+3. Copy the client ID, we will need it in the following steps.
+
+## Add user in Casdoor
+
+Now you have the application running, but you must create a user and assign them a role next.
+
+Go to the “Users” page and click “Add user” in the top right corner. This will open a new page where you can add the
+user.
+
+![Pic showing the users page](https://github.com/RanTao123/image/blob/main/user.png?raw=true)
+
+Save the user after adding a username and the organization Hasura (other details are optional).
+
+Now you need to set up a password for your user, which you can do by clicking "Manage Your Password."
+
+Choose a password for your user and confirm it.
+
+## Create a Hasura instance
+
+Create a Hasura instance using either
+[Hasura Cloud](https://hasura.io/docs/latest/getting-started/getting-started-cloud/) or
+[Docker](https://hasura.io/docs/latest/getting-started/docker-simple/).
+
+Now, create a `users` table with the following columns:
+
+- `id` of type Text (Primary Key)
+- `username` of type Text
+
+See the image below for reference.
+
+![Picture showing how to create a table in Hasura](https://graphql-engine-cdn.hasura.io/learn-hasura/assets/graphql-hasura-authentication/keycloak/hasura-create-table.png)
+
+The next step is to create a `user` role for the app. Users should be able to see only their records, but not the other
+people’s records.
+
+Configure the `user` role as shown in the image below. For more information, read about
+[configuring permission rules in Hasura](https://hasura.io/docs/latest/graphql/core/auth/authorization/permission-rules/).
```

Comment:
Again, this isn’t clear enough for a user. You need to provide more detail explaining how a user should configure the permissions for the user role. Specifically, are we configuring a permission rule on the `id` or `username` field?